### DISCUSSION TITLE
Hierarchical data structure question

### DISCUSSION THREAD

**Discussion by JezhikLaas** [ravendb/ravendb#15176] [Q&A] [answered]
Hello,

we have (basically) the following data structure:
`class Test
{

    public string Value { get; set; }

    public string ParentValue { get; set; }

    ...
}`

So there is a parent / child relationship defined by the property 'ParentValue'. We do not maintain a list of children within a parent, because a parent node would then be updated concurrently.

We do want to query a hierarchy starting with a parent node and want to get all its children, grandchildren and so on. We fail to implement this without querying the single levels recursively. Any ideas how to improve this? It would be possible for us to use the document id as the link to a parent, but I do not see, how this could help.

Thanks, Uwe

**Comment by ayende** [comment, marked as answer]
What is the depth that you are likely looking at?

In general, the easy way to do that is to walk the parents on indexing, so you can do something like:

```
map("Employees", e => {
   var bosses = [];
    var bossId = e.Boss;
   while(bossId != null){
     bosses.push(bossId)
      var boss = load(bossId, "Employees");
      bossId = boss.Boss;
   }
  return { Bosses: bosses }
});
```

The issue with that is that whenever the CEO changes, _all_ employees would have to be indexed to reflect that.

Alternatively, you can put a limit (1 - 2 levels) and see if that is actually something that you need.

What sort of queries do you actually need to run?

**Reply by JezhikLaas** [reply]
Hello Oren,
the data structure models production units (0) aggregated into the levels trade unit (1), ship carton (2), pallet (4) and we want get all units which are aggregated into a certain level, for example: which units are in ship carton XYZ. Maximum level for this query would be if we query from a pallet which carries around 5000 units at max. This would be 4 levels, maybe we get to 5 later on when we gather pallets into deliveries, but there won't be more then that.
Having an index as you suggested would make it much easier for us to use the streaming API.

I will give your suggestion a try an set the answer flag to this question if this works well.

Thanks, Uwe

**Reply by JezhikLaas** [reply]
Works perfectly, thank you!

**Reply by ayende** [reply]
Do you make modifications to the production units afterward?
If this is 500 total in depth, then the index is really a great option ,yes.

**Reply by JezhikLaas** [reply]
It is possible, but rather seldom that a bundle (trade unit, carton or pallet) gets destroyed and the children re-aggregated into a new parent. The overall amount of production units will be several millions over the next years - this is an industrial production of small units (several dozens per second) which we have to track one by one due to EU regulations. But the parent / child tree will never be deeper than 5. It is an interesting project and Raven really shines so far.

**Comment by mcrio** [comment]
Hi @JezhikLaas out of curiosity what does your index code look like? If I understand correctly each Index entry now contains an array of children? Wondering if the index will grow very big that way...

**Reply by JezhikLaas** [reply]
Hi,
for now I have only a test setup to see, if it works like we need it to:
`map('RecursiveTests', e => {
    var parents = [];
    var parentId = e.ParentId;
    while (parentId != null) {
        parents.push(parentId);
        var parent = load(parentId, 'RecursiveTests');
        parentId = parent.ParentId;
    }
    return { Parents: parents };
});`

Very similar to Orens suggestion. But yes, you are right, every node in this tree produces a bunch of entries in the respective index. The leafs (the single production units) will have three entries, the next level 2 and so on.

**Reply by JezhikLaas** [reply]
Additional note: we expect the system to handle 86 million production units per year.
